### DISCUSSION TITLE
function of a variable

### DISCUSSION THREAD

**Discussion by tqcthai** [idaholab/moose#25812] [Q&A Modules: General] [answered]
Dear all,

I know that functions in moose depend only on spatial position and time: f(x,y,z,t). How can I define a function of a variable?
I need to use it for AMR and this way doesn't work.
``
````
[Variables]
  [c]
  order = FIRST
  family = LAGRANGE
  []
[]

[Functions]
  [I1]
  type = ParsedFunction
  expression = min(c,0.1)
  []
[]

[Adaptivity]
  marker = errorfrac
  steps = 2
  max_h_level = 5
  [Indicators]
    [error]
      type = AnalyticalIndicator
      variable = c
      function = I1
     #function = min(c, 0.1)
    []
  []
  [Markers]
    [errorfrac]
      type = ErrorFractionMarker
      coarsen = 0.0
      indicator = error
      refine = 1e-7
    []
  []
[]`                  
```

**Comment by GiudGiud** [comment]
Hello

It likely should be a material property

For output only, you can also use an auxiliary variable.

Guillaume

**Reply by tqcthai** [reply]
Hi GuidGuid,

I' ve tried

```
[Functions]
  [I1]
  type = ParsedFunction
  expression = min(elastic_energy,0.1)
  []
[]

[Adaptivity]
  marker = errorfrac
  steps = 2
  max_h_level = 4
  [Indicators]
    [error]
      type = AnalyticalIndicator
      variable = c
      function = I1
    []
  []
  [Markers]
    [errorfrac]
      type = ErrorFractionMarker
      coarsen = 0.0
      indicator = error
      refine = 1e-7
    []
  []
[]`

```
where elastic_energy is a material property. This is the error message I got:
ERROR: FunctionParser is unable to parse expression: min(elastic_energy,0.1)
Syntax error: Unknown identifier

**Reply by GiudGiud** [reply]
cant do that. A Parsed function can only be dependent on x y z t, postprocessors and other functions

you ll have to use a material property or an auxkernel (that fills an auxvariable) to do this operation.
An aux-variable can be used as an indicator

**Reply by tqcthai** [reply]
Hello,

I made it like this

```
[Adaptivity]
  marker = errorfrac
  max_h_level = 4
  [Indicators]
    [error]
      type = AnalyticalIndicator
      variable = 'c_indicator'
      function = '0'
    []
  []
  [Markers]
    [errorfrac]
      type = ErrorFractionMarker
       indicator = error
      refine = 1e-3
    []
  []
[]

[AuxVariables]
  [c_indicator]
  []
  [resid_x]
  []
[]
[Material]
  ....
  output_properties = 'c_indicator'
[]
[Modules/TensorMechanics/Master]
  [all]
      ...
     save_in = 'resid_x'
  []
[]
```
I specified function of c_indicator inside the material file. Although there is no error in running, the value of c_indicator in the output is always zero. When I change the indicator as resid_x which is from the tensor mechanics module output, it works. I wonder how to make the output of c_indicator from materials block can be recognized as resid_x?

**Reply by GiudGiud** [reply, marked as answer]
Use a MaterialRealAux to move a material property to an auxiliary variable.

what you have now is designed for the output to Exodus file of matrial properties

**Reply by tqcthai** [reply]
It works. Thank you so much.
